src/numerics/Newton.cpp

```diff
@@ -0,0 +1,339 @@
+    // m_atol_ts.resize(m_nv, 1.0e-11);
+
+    m_jacobian = DenseMatrix(m_nv, m_nv);
+    m_jacAge = 10000;
+    m_jacMaxAge = 5;
+    m_jacRtol = 1.0e-5;
+    m_jacAtol = sqrt(std::numeric_limits<double>::epsilon());
+}
+
+void Newton::evalJacobian(doublereal* x, doublereal* xdot) {
+
+    // // calculate unperturbed residual
+    m_residfunc->eval(0, x, xdot, 0);
+
+    for (size_t n = 0; n < m_nv; n++) {
+        // calculate the nth Jacobian column, unless component n is constant
+        if (!m_constant[n]) {
+            double xsave = x[n];
+
+            // calculate the perturbation amount, preserving the sign of x[n]
+            double dx;
+            if (xsave >= 0) {
+                dx = xsave*m_jacRtol + m_jacAtol;
+            } else {
+                dx = xsave*m_jacRtol - m_jacAtol;
+            }
+
+            // perturb the solution vector
+            x[n] = xsave + dx;
+            dx = x[n] - xsave;
+
+            // calculate perturbed residual
+            vector_fp xdotPerturbed(m_nv); //make this member for speed?
+            m_residfunc->eval(0, x, xdotPerturbed.data(), 0);
+
+            // compute nth column of Jacobian
+            for (size_t m = 0; m < m_nv; m++) {
+                m_jacobian.value(m,n) = (xdotPerturbed[m] - xdot[m])/dx;
+            }
+            // restore solution vector
+            x[n] = xsave;
+
+        // for constant components: Jacobian column of 0's, with 1 on diagonal
+        // note: Jacobian row must also be 0's w/ 1 on diagonal
+        } else {
+            for (size_t m = 0; m < m_nv; m++) {
+                m_jacobian.value(m,n) = 0;
+            }
+            m_jacobian.value(n,n) = 1;
+        }
+    }
+
+    // writelog("\nnew jac:\n");
+    // for (int i = 0; i < m_nv; i++) {
+    //     for (int j = 0; j < m_nv; j++) {
+    //         writelog("{:14.5} ", m_jacobian.value(i,j));
+    //     }
+    //     writelog("\n");
+    // }
+
+    m_jacAge = 0;
+}
+
+doublereal Newton::weightedNorm(const doublereal* x, const doublereal* step) const
+{
+    double sum = 0.0;
+    for (size_t n = 0; n < m_nv; n++) {
+        double weight = m_rtol_ss[n]*fabs(x[n]) + m_atol_ss[n]; //TODO: add transient tolerances if rdt!=0
+        double f = step[n]/weight;
+        sum += f*f;
+    }
+    return sqrt(sum/m_nv);
+}
+
+void Newton::step(doublereal* x, doublereal* step, int loglevel)
+{
+    m_residfunc->eval(0, x, step, 0);
+    for (size_t n = 0; n < m_nv; n++) {
+        step[n] = -step[n];
+    }
+
+    DenseMatrix solvejac = m_jacobian;
+    try {
+        //Note: this function takes an unfactored jacobian, then finds its LU factorization before
+        // solving. Optimization is possible by saving the factored jacobian, since it can be reused.
+        // Also, the DenseMatrix provided here will be overwritten with the LU factored version, so
+        // a copy is passed instead in order to preserve the original for reuse.
+        Cantera::solve(solvejac, step);
```

Comment:
This optimization would need to be made in `DenseMatrix`-- two options:
1. Add `factor()` and `solveFactored()` methods to `DenseMatrix`, keeping the factored matrix in `Newton`
2. Add `m_factored` and `m_factoredData` to `DenseMatrix`, to store and keep track of a factored matrix, then choose between the factored/unfactored versions in `solve()` based on `m_factored`